### PR TITLE
[tune] Conditional search space example using hyperopt

### PR DESCRIPTION
**Pull request by ryanlmelvin**
## Why are these changes needed?

Provides an example of conditional search spaces.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — CHANGES_REQUESTED** [review]
Awesome, thanks! Could you also update the docs to include the example and ensure it runs in CI? You can follow this PR as a guide: https://github.com/ray-project/ray/pull/17464 

Also, please run the linting script in `scripts/format.sh` to make Lint CI pass!

**Comment by ryanlmelvin** [comment]
@Yard1 I think I've made the changes you wanted: conf and doc files updated. CI ran without error locally. Please review.

File: doc/source/tune/examples/hyperopt_conditional_search_space_example.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+:orphan:
+
+hyperopt_conditional_search_space_example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (Yard1) [inline doc/source/tune/examples/hyperopt_conditional_search_space_example.rst:4]:
```suggestion
hyperopt_conditional_search_space_example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

File: doc/source/tune/examples/hyperopt_conditional_search_space_example.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+:orphan:
+
+hyperopt_conditional_search_space_example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (Yard1) [inline doc/source/tune/examples/hyperopt_conditional_search_space_example.rst:4]:
```suggestion
hyperopt_conditional_search_space_example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

**Review by Yard1 — APPROVED** [review]
Yep, looks good! Waiting for CI to pass.

File: python/ray/tune/examples/hyperopt_conditional_search_space_example.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    
+    Source: https://www.kaggle.com/fanvacoolt/tutorial-on-hyperopt
+    
+    Parameters:
+    ----------------
+    dct : dictionary to unpack
+    
+    Returns:
+    ----------------
+    : unpacked dictionary
```
Comment (Yard1) [inline python/ray/tune/examples/hyperopt_conditional_search_space_example.py:32]:
```suggestion
    Unpacks all sub-dictionaries in given dictionary recursively.
    There should be no duplicated keys across all nested
    subdictionaries, or some instances will be lost without warning

    Source: https://www.kaggle.com/fanvacoolt/tutorial-on-hyperopt

    Parameters:
    ----------------
    dct : dictionary to unpack

    Returns:
    ----------------
    : unpacked dictionary
```

**Review by krfricke — APPROVED** [review]
Thanks, this is great!

**Pull request merged**